jax/_src/lax/control_flow/loops.py

```diff
@@ -263,106 +286,197 @@ def _create_jaxpr(init):
 
   _check_tree_and_avals("scan carry output and input",
                         # Extract the subtree and avals for the first element of the return tuple
-                        out_tree_children[0], carry_avals_out,
+                        out_tree_children[1], carry_avals_out,
                         init_tree, carry_avals)
   disallowed_effects = jaxpr.effects - allowed_effects
   if disallowed_effects:
     raise NotImplementedError(
         f'Effects not supported in `scan`: {disallowed_effects}')
 
-  out = scan_p.bind(*consts, *in_flat,
+  # use default precision with no weak types
+  start = jax.numpy.array([0])[0]
+  stop = jax.numpy.array([length])[0]
+  done = jax.numpy.array(early_exit(init), dtype=np.bool_)
```

Comment:
I suspect there's a better way to accomplish this but I don't know what it is.

Reply:
Besides introducing the extra `done` variable, we also introduce `start` and `stop`. The iteration happens over 'xs[start:stop]`. (In particular inclusive of `start` but exclusive of `stop`.) This is needed when backpropagating through an early-exit'd scan: we would like to start from partway through the backward iterations.